workbench/src/wb_project.c

```diff
@@ -0,0 +1,1491 @@
+/*
+ * Copyright 2017 LarsGit223
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+
+/*
+ * Code for the WB_PROJECT structure.
+ */
+#include <glib/gstdio.h>
+
+#ifdef HAVE_CONFIG_H
+# include "config.h"
+#endif
+
+#include <geanyplugin.h>
+#include "wb_globals.h"
+#include "wb_project.h"
+#include "utils.h"
+
+extern GeanyData *geany_data;
+
+typedef enum
+{
+	WB_PROJECT_TAG_PREFS_AUTO,
+	WB_PROJECT_TAG_PREFS_YES,
+	WB_PROJECT_TAG_PREFS_NO,
+}WB_PROJECT_TAG_PREFS;
+
+typedef struct
+{
+	GKeyFile *kf;
+	guint    dir_count;
+}WB_PROJECT_ON_SAVE_USER_DATA;
+
+struct S_WB_PROJECT_DIR
+{
+	gchar *name;
+	gchar *base_dir;
+	gchar **file_patterns;	/**< Array of filename extension patterns. */
+	gchar **ignored_dirs_patterns;
+	gchar **ignored_file_patterns;
+	guint file_count;
+	guint folder_count;
+	GHashTable *file_table; /* contains all file names within base_dir, maps file_name->TMSourceFile */
+};
+
+struct S_WB_PROJECT
+{
+	gchar     *filename;
+	gchar     *name;
+	gboolean  modified;
+	GSList    *s_idle_add_funcs;
+	GSList    *s_idle_remove_funcs;
+	GSList    *directories;  /* list of WB_PROJECT_DIR; */
+	WB_PROJECT_TAG_PREFS generate_tag_prefs;
+	GPtrArray *bookmarks;
+};
+
+
+/** Set the projects modified marker.
+ *
+ * @param prj   The project
+ * @param value The value to set
+ *
+ **/
+void wb_project_set_modified(WB_PROJECT *prj, gboolean value)
+{
+	if (prj != NULL)
+	{
+		prj->modified = value;
+	}
+}
+
+
+/** Has the project been modified since the last save?
+ *
+ * @param prj   The project
+ * @return TRUE if project is modified, FALSE otherwise
+ *
+ **/
+gboolean wb_project_is_modified(WB_PROJECT *prj)
+{
+	if (prj != NULL)
+	{
+		return (prj->modified);
+	}
+	return FALSE;
+}
+
+
+/** Set the filename of a project.
+ *
+ * @param prj      The project
+ * @param filename The filename
+ *
+ **/
+void wb_project_set_filename(WB_PROJECT *prj, gchar *filename)
+{
+	if (prj != NULL)
+	{
+		guint offset;
+		gchar *ext;
+
+		prj->filename = g_strdup(filename);
+		prj->name = g_path_get_basename (filename);
```

Comment:
Well, today the function is only called once. But I changed it to be on the safe side in case of future changes.